Initial implementation of SMILES Django Portal

lahirujayathilake opened the pull request:
- Protobuf schemas for SMILES data products
- DataCatalog python client implementation with the required CRUD operations for the data catalog models
- Retrieve all the DPs per given type combining multiple schemas
- Celery and Redis-based solution for handling large file uploads
- Vue Js development for File Uploads, visualize the different types of data products with search, experimental data detailed view

machristie commented:
@lahirujayathilake is this ready to be reviewed? Just want to check before I start reviewing.

lahirujayathilake commented:
Hi @machristie, due to some recent changes, there are a few parts that I need to complete. I will finish them and let you know when this is ready for review.

lahirujayathilake commented:
@machristie this is ready for review

machristie commented on frontend/src/components/DataProductTable.vue:
```diff
@@ -0,0 +1,184 @@
+          });
+    },
+    edit(data) {
+      alert(JSON.stringify(data));
```
Not sure what edit should do here, but let's not use `alert()`. Use a modal or some other BootstrapVue component if you need a popup window.

lahirujayathilake replied:
The edit functionality will be removed from this Vue component and instead, it will be added to the detailed page

machristie commented on frontend/src/components/DataProductUploader.vue:
```diff
@@ -0,0 +1,318 @@
+
+  computed: {
+    uploadUrl() {
+      const URL_PREFIX = 'http://127.0.0.1:8000/smiles/';
```
Change this to `'/smiles/'`. With the webpack loader integration, you won't need this any longer. Also, you shouldn't need any special CORS configuration.

machristie commented on frontend/src/services/configuraion-service.js:
```diff
@@ -0,0 +1,40 @@
+
+function getDisplayableColumns(type) {
+    if (type === "exp") {
+        return JSON.parse(localStorage.getItem('exp-display-columns')) ||
```
Question: what is localStorage doing here? I looked for code that was writing to localStorage but didn't find it. I'm wondering if this is needed any longer.

lahirujayathilake replied:
Making the addition/removal of columns configurable based on user preferences is a requirement for the SMILES project. Therefore, I have started using local storage for this feature. 
The remaining part of the functionality will be implemented within this project.

machristie commented on frontend/src/services/smiles-dp-service.js:
```diff
@@ -0,0 +1,26 @@
+
+function getSMILESDataProduct(type, dp_id) {
+    return axios
+        .get(`http://127.0.0.1:8000/smiles/${type}-dp/${dp_id}`)
```
Likewise in this file, the url can start with `/smiles/`

machristie commented on smiles/views.py:
```diff
@@ -0,0 +1,195 @@
+from .apps import SmilesDjangoPortalConfig
+
+
+@method_decorator(csrf_exempt, name='dispatch')
```
This is fine for now but it would be better to include the CSRF token in REST requests. See https://docs.djangoproject.com/en/4.2/howto/csrf/#acquiring-the-token-if-csrf-use-sessions-and-csrf-cookie-httponly-are-false

machristie reviewed (changes requested):
This looks really good @lahirujayathilake ! Nice job. I've left some review comments.

lahirujayathilake commented:
Thanks, @machristie! I really appreciate you reviewing the code and providing feedback on my work. 
I've addressed all the review comments in the latest commit.

machristie commented:
Looks good @lahirujayathilake , thanks
